[üîù](../README.md)

# 24/04/2024 - Aula 019 -

Prof. Perrotti
POO

## Thread

√â uma linha de processamento dentro do programa. Dentro de uma thread, os comandos s√£o executados em sequ√™ncia, um comando s√≥ vai inicar sua execu√ß√£o quando o anterior dor finalizado. O programa pode abrir v√°rios threads, nesse caso, elas ser√£o executadas em paralelo, uma independente das outras. Mas pelomentos uma thread √© obrigat√≥ria a thread principal que comt√©m a fun√ß√£o main().

## Pilha de execu√ß√£o (stack)

Para cada thread aberta pelo programa, √© criada uma √°rea de mem√≥ria para conter os dados das fun√ß√µes que est√£o em execu√ß√£o no momento (stack). Nessa √°rea fica armazenados os dados das fun√ß√µes em execu√ß√£o.

Tudo o que for declarado em uma fun√ß√£o, como vari√°veis em parametros, ficam armazenados nessa √°rea. Sempre que uma fun√ß√£o √© chamada outra fun√ß√£o, a que fez a chamada entra em estado de espera e √© alocada uma por√ß√£o de mem√≥ria logo acima no stack para a fun√ß√£o que foi chamada.

## Execu√ß√£o

√â gerada quando uma situa√ß√£o inv√°lida impede o processamento normal do programa. A exe√ß√£o √© sempre lan√ßada na fun√ß√£o que estiver no topo da pilha, j√° que estiver no topo da pilha, j√° que √© a √∫nica que est√° realmente executando.

O objetivo da exe√ß√£o √© notificar a fun√ß√£o que uma situa√ß√£o de erro est√° impedindo o processamento. A fun√ß√£o que recebe a exece√ß√£p pode capturar ou ignorar. Se capturar, a exce√ß√£o √© destru√≠da e o processamento segue normalmente.

Mas se aexe√ß√£o for ignorada ( ou seja n√£o capturada) a fun√ß√£o ser√° abortada e a exce√ß√£o. Se nenhuma fun√ß√£o capturar a exe√ß√£o na fun√ß√£o main() que tamb√©m ser√° abortada caso n√£o a capture, abortando o programa.

## try - catch - finally

Bloco `try` -> √â onde fica o c√≥digo que est√° sugeito a exce√ß√µes. Vai ser executado at√© o ponto onde a exce√ß√£o ocorrer.

Bloco `catch(NomeExecao exc)` -> √â o bloco que captura e trata as exe√ß√µes. Pode haver v√°rios blocos catch, um para cada exce√ß√£o que ser√° capturada. Apenas um bloco catch ser√° executado.

Bloco `finally` -> Este bloco sempre ser√° executado, ocorrendo ou n√£o a exce√ß√£o. Use para finalizar opre√ß√µes que n√£o podem ficar pendentes.

## Classe Exception

√â ancestral de todas as exce√ß√µes no java.